docs/internal/capabilities.md

```diff
@@ -0,0 +1,210 @@
+* Arena model: Objects are allocated and used, but only freed on the destruction of the region itself.
+
+They can also refer to how the code is executed:
+* Native: Verona code calling Verona code, fully checked and safe.
+* Sandbox: Verona code calling an external language module (ex. a C++ library), with process isolation.
+
+Sandboxes themselves can also be of different types, for example:
+* Child process: Forks a new process, shared memory, calls through RPC.
+* WebASM: Isolates OS functionality, memory access, etc.
+* Remote execution: For use with different technologies (ex. GPU) or architectures (ex. connected embedded board).
+
+None of those lists are complete, they're just examples of what can be done.
+
+## Assignment
+
+Capabilities change the semantics of assignments in the language.
+
+**Isolated objects** can only be used through a single reference, so assigning an `iso` region to a new variable moves the reference and invalidates the previous one.
+
+```ts
+var foo : Foo & iso = Foo; // plus undecided syntax for "in new region type X"
+foo.doSomething();
+...
+var bar = foo; // bar now holds the *only* reference to foo
+bar.somethingElse(); // OK: bar is a reference to the same object as foo
+foo.cleanup(); // ERROR: foo is not a valid reference any more
+```
+
+**Mutable objects** can alias each other, so assigning a `mut` variable to another creates a (different) reference to the same variable.
+Given that mutable references can only be modified by the one behaviour that holds their dominating `iso` reference, there are no race conditions in any write or reassignment.
+
+```ts
+var foo : Foo & mut = Foo::create(10); // Returns a Foo object with value = 10
+print(foo.value); // prints "10"
+...
+// Binding a new variable to the same reference creates a new reference to the same object
+var bar = foo;
+bar.value = 42;
+print(foo.value); // prints "42"
+...
+// Creating a new object reassigns the variable bar but doesn't change the previous references (foo)
+bar = Foo::create(31); // Returns a new Foo object with value = 31, binds it to bar
+print(bar.value); // prints "31"
+// But foo still points to the previous object
+print(foo.value); // still prints "42"
+```
+
+**Read-only references** can be created from other capabilities, but cannot be assigned back to mutable references later.
+
+This is to protect passing a read-only reference and the user change it back to a mutable one.
+```ts
+var foo : Foo & mut = Foo::create(10); // Returns a Foo object with value = 10
+print(foo.value); // prints "10"
+...
+var bar : Foo & readonly = foo; // OK: a read-only reference to a mutable object
+...
+var baz : Foo & mut = bar; // ERROR: cannot convert readonly to mutable
+```
+
+**Immutable objects** can also alias each other, but since none of those references can write through them, once created (frozen), an immutable region can only be read from.
```

Comment:
```suggestion
**Immutable capabilities enforce global immutability**.
The `freeze` operation consumes an `iso` capability.
The type system therefore guarantees that there can be no pointers to objects in the region (or any child regions) anywhere other than as pointers within the regions being frozen.
The `freeze` operation returns an `imm` capability and *viewpoint adaptation* guarantees that this cannot be used to materialise any kind of mutable capability.
Once a region has been frozen, no capabilities exist that permit writes to any objects that were in that region.
```

Reply:
Should we explain or link to a page explaining "viewpoint adaptation"? The first hits on search engines are all about computer vision and object detection, which can be super hard to filter.

Also, the Pony tutorial talks about viewpoint adaptation, but the Pony types are so much more complex that I'm not sure how we'd relate in the same way.